src/server/auth/server/api_server.go

```diff
@@ -1416,22 +1354,157 @@ func (a *apiServer) RestoreAuthToken(ctx context.Context, req *auth.RestoreAuthT
 		}
 	}
 
-	// Check whether the token hash already exists - we don't want to replace an existing token
-	if _, err := col.NewSTM(ctx, a.env.GetEtcdClient(), func(stm col.STM) error {
-		tokens := a.tokens.ReadWrite(stm)
-		var existing auth.TokenInfo
-		err := tokens.Get(req.Token.HashedToken, &existing)
-		if err == nil {
-			return errors.New("cannot overwrite existing token with same hash")
-		} else if err != nil && !col.IsErrNotFound(err) {
-			return err
+	if err := func() error {
+		if ttl > 0 {
+			return a.insertAuthToken(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject, ttl)
+		} else {
+			return a.insertAuthTokenNoTTL(ctx, req.Token.HashedToken, req.Token.TokenInfo.Subject)
 		}
-
-		return tokens.PutTTL(req.Token.HashedToken,
-			req.Token.TokenInfo,
-			ttl)
-	}); err != nil {
+	}(); err != nil {
 		return nil, errors.Wrapf(err, "error restoring auth token")
 	}
+
 	return &auth.RestoreAuthTokenResponse{}, nil
 }
+
+// implements the protobuf auth.DeleteExpiredAuthTokens RPC
+func (a *apiServer) DeleteExpiredAuthTokens(ctx context.Context, req *auth.DeleteExpiredAuthTokensRequest) (*auth.DeleteExpiredAuthTokensResponse, error) {
+	logrus.Info("Deleting expired auth tokens")
+	if _, err := a.env.GetDBClient().Exec(`DELETE FROM auth.auth_tokens WHERE NOW() > expiration`); err != nil {
+		return nil, errors.Wrapf(err, "error deleting expired tokens")
+	}
+	return &auth.DeleteExpiredAuthTokensResponse{}, nil
+}
+
+func (a *apiServer) deleteExpiredTokensRoutine() {
+	go func(ctx context.Context) {
+		for {
+			time.Sleep(time.Duration(cleanupIntervalHours) * time.Hour)
+			a.DeleteExpiredAuthTokens(ctx, &auth.DeleteExpiredAuthTokensRequest{})
+		}
+	}(context.Background())
+}
+
+// we interpret an expiration value of NULL as "lives forever".
+func (a *apiServer) lookupAuthTokenInfo(ctx context.Context, tokenHash string) (*auth.TokenInfo, error) {
+	var subject string
+	var expiration *time.Time
+
+	rows, err := a.env.GetDBClient().QueryxContext(ctx, `SELECT subject, expiration FROM auth.auth_tokens WHERE token_hash = $1`, tokenHash)
+	if err != nil || !rows.Next() {
+		return nil, col.ErrNotFound{Type: "auth_tokens", Key: tokenHash}
+	}
+
+	if err = rows.Scan(&subject, &expiration); err != nil {
+		return nil, errors.Wrapf(err, "error querying token")
+	}
+
+	tokenInfo := auth.TokenInfo{Subject: subject}
+	if expiration != nil {
+		expiryTS, tsErr := types.TimestampProto(*expiration)
```

Comment:
There's a gogoproto extension that will generate this struct with a normal Go time, which should let you load the data directly into the struct using sqlx:

https://github.com/pachyderm/pachyderm/blob/master/src/license/license.proto#L61-L62

https://github.com/pachyderm/pachyderm/blob/master/src/server/license/server/api_server.go#L308-L309